### PR TITLE
Implement DataPipe to convert Iter to Map

### PR DESCRIPTION
**Pull request by ejguan**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #202 Revamp librispeech dataset
* #203 [DataPipe] Fix OnDiskCacheHolder to list all files for decompressing operations
* **#201 Implement DataPipe to convert Iter to Map**

This DataPipe is used to lazily load data from `IterDataPipe` to a `MapDataPipe`.

Differential Revision: [D34085744](https://our.internmc.facebook.com/intern/diff/D34085744)

### CODE REVIEW

File: torchdata/datapipes/map/util/utils.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+            self.key_fn = dill.loads(dill_key_fn)  # type: ignore[assignment]
+            self.value_fn = dill.loads(dill_value_fn)  # type: ignore[assignment]
+        else:
+            self.key_fn = dill_key_fn  # type: ignore[assignment]
+            self.value_fn = dill_value_fn  # type: ignore[assignment]
+
+
+# Register for functional API
+# See Issue https://github.com/pytorch/data/issues/200
+IterDataPipe.register_datapipe_as_function("to_map", IterToMapConverterMapDataPipe)
```
Comment (ejguan) [inline torchdata/datapipes/map/util/utils.py:98]: Note

**Comment by ejguan** [comment]
@ejguan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34085744).

**Review by NivekT — APPROVED** [review]
LGTM with a few comments.

File: torchdata/datapipes/map/util/utils.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+if DILL_AVAILABLE:
+    import dill
+
+    dill.extend(use_dill=False)
+
+
+class IterToMapConverterMapDataPipe(MapDataPipe):
+    r"""
+    Lazily load data from IterDataPipe to construct a dictionary with the
+    key-value pair generated by ``key_fn`` and ``value_fn``.
```
[the remark this answers is not in the document]
Comment (NivekT) [reply torchdata/datapipes/map/util/utils.py:18]: We can manually add this DataPipe to `torchdata.datapipes.map.rst` and add the functional API to the end of the first sentence

[the remark this answers is not in the document]
Comment (ejguan) [reply torchdata/datapipes/map/util/utils.py:18]: But, functional API should also be split into `map` and `iter` right? Then, how should we differentiate them?

File: torchdata/datapipes/map/util/utils.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+            dill_key_fn = dill.dumps(self.key_fn)
+            dill_value_fn = dill.dumps(self.value_fn)
+        else:
+            dill_key_fn = self.key_fn
+            dill_value_fn = self.value_fn
+        return (
+            self.datapipe,
+            dill_key_fn,
+            dill_value_fn,
+            self._map,
```
[the remark this answers is not in the document]
Comment (ejguan) [reply torchdata/datapipes/map/util/utils.py:88]: Good catch. Adding it.

File: torchdata/datapipes/map/util/utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (NivekT) [reply torchdata/datapipes/map/util/utils.py:18]: I see what you mean. I think these are the options:

1. Put this DataPipe in `torchdata.datapipes.iter.rst` and have label `to_map_datapipe` as functional API. The logic is that this DataPipe takes in a `IterDataPipe` and returns a `MapDataPipe` and the functional API only works on `IterDataPipe`.

2. Have a separate documentation page that discusses different converters (e.g. to TorchArrow DataFrame, to `MapDataPipe`, etc).

Do you have other suggestion?

[the remark this answers is not in the document]
Comment (ejguan) [reply torchdata/datapipes/map/util/utils.py:18]: I see. Then, I will go for the first one. I will add it into `torchdata.datapipes.iter.rst`

**Comment by VitalyFedyunin** [comment]
Besides Kevin's comments, I find it is little bit confusing to have `.map` and `.to_map`. Should we force `.to_map` arguments to be kwargs only? Maybe `.to_map(key_fn=x, value_fn=x)` will be cleaner. I'm open to suggestions.

**Comment by ejguan** [comment]
> Besides Kevin's comments, I find it is little bit confusing to have `.map` and `.to_map`. Should we force `.to_map` arguments to be kwargs only? Maybe `.to_map(key_fn=x, value_fn=x)` will be cleaner. I'm open to suggestions.

Do you think it's still confusing if we change it to `.to_map_datapipe`?

**Comment by VitalyFedyunin** [comment]
I'm fine with `.to_map_datapipe`

**Comment by ejguan** [comment]
@ejguan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34085744).

**Comment by ejguan** [comment]
@ejguan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34085744).

**Comment by ejguan** [comment]
@ejguan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34085744).

**Pull request closed without merging**